staging/src/k8s.io/api/autoscaling/v2beta2/types.go

```diff
@@ -0,0 +1,361 @@
+	Target          MetricTarget                `json:"target" protobuf:"bytes,2,name=target"`
+	Metric          MetricIdent                 `json:"metric" protobuf:"bytes,3,name=metric"`
+}
+
+// PodsMetricSource indicates how to scale on a metric describing each pod in
+// the current scale target (for example, transactions-processed-per-second).
+// The values will be averaged together before being compared to the target
+// value.
+type PodsMetricSource struct {
+	Metric MetricIdent  `json:"metric" protobuf:"bytes,1,name=metric"`
+	Target MetricTarget `json:"target" protobuf:"bytes,2,name=target"`
+}
+
+// ResourceMetricSource indicates how to scale on a resource metric known to
+// Kubernetes, as specified in requests and limits, describing each pod in the
+// current scale target (e.g. CPU or memory).  The values will be averaged
+// together before being compared to the target.  Such metrics are built in to
+// Kubernetes, and have special scaling options on top of those available to
+// normal per-pod metrics using the "pods" source.  Only one "target" type
+// should be set.
+type ResourceMetricSource struct {
+	// Name is the name of the resource in question.
+	Name   v1.ResourceName `json:"name" protobuf:"bytes,1,name=name"`
+	Target MetricTarget    `json:"target" protobuf:"bytes,2,name=target"`
+}
+
+// ExternalMetricSource indicates how to scale on a metric not associated with
+// any Kubernetes object (for example length of queue in cloud
+// messaging service, or QPS from loadbalancer running outside of cluster).
+type ExternalMetricSource struct {
+	Metric MetricIdent  `json:"metric" protobuf:"bytes,1,name=metric"`
+	Target MetricTarget `json:"metricTarget" protobuf:"bytes,2,name=metricTarget"`
+}
+
+// MetricIdent defines the name and optionally selector for a metric
+type MetricIdent struct {
+	// Name is the name of the given metric
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+	// selector is the selector for the given metric
+	// it is the string-encoded form of a standard kubernetes label selector
+	// +optional
+	Selector string `json:"selector,omitempty" protobuf:"bytes,2,name=selector"`
+}
+
+// MetricTarget defines the target value, average value, or average utilization of a specific metric
+type MetricTarget struct {
+	// Type represents whether the metric type is Utilization, Value, or AverageValue
+	Type         MetricTargetType   `json:"type" protobuf:"bytes,1,name=type"`
+	Value        *resource.Quantity `json:"value,omitempty" protobuf:"bytes,2,name=value"`
+	AverageValue *resource.Quantity `json:"averageValue,omitempty" protobuf:"bytes,3,name=averageValue"`
+
+	// Currently only valid for Resource metric source type
+	AverageUtilization *Percentage `json:"averageUtilization,omitempty" protobuf:"bytes,4,name=averageUtilization"`
+}
+
+// Percentage is an integer internally (like before), but
+// serializes as "<number>%" to avoid user confusion.
+type Percentage = int32
```

Comment:
This is clever and novel, but I'm not sold.  Where is the serialization actually defined?  What happens if the user omits the percentage?  What if they provide `0`?   Is `%` by itself allowed?  How are you reporting errors back to the user.

I'm going to need some more justification before we introduce something like this.  And it will need a lot more documentation.

Reply:
I don't think the code every got implemented for this -- it was supposed to be yanked.  I requested API review on it because I wasn't sure if it was legit.

My hope with the original proposal was that it would make it clearer to users what this field was, and what the units were, because as it stands, it's a bit confusing.  It's no big deal with it gets removed, but I would like to consider it eventually.